### PR TITLE
Update the `PT_XLA_DEBUG` to make it recognize the `torch_xla.compile`

### PR DESCRIPTION
**Pull request by JackCaoG**
the expected output looks like
```
Compilation Analysis: ================================================================================
Compilation Analysis: Compilation Cause
Compilation Analysis:   torch_xla.compile clear the pending graph prior calling the target function
Compilation Analysis: Graph Info: 
Compilation Analysis:   Graph Hash: 1abebf4362b8342721a64502bd4074c1
Compilation Analysis:   Number of Graph Inputs: 1
Compilation Analysis:   Number of Graph Outputs: 1
Compilation Analysis: Python Frame Triggered Execution: 
Compilation Analysis:   mark_step (/workspaces/dk4/pytorch/xla/torch_xla/core/xla_model.py:1008)
Compilation Analysis:   _compile (/workspaces/dk4/pytorch/xla/torch_xla/torch_xla.py:108)
Compilation Analysis:   __enter__ (/usr/local/lib/python3.10/contextlib.py:135)
Compilation Analysis:   inner (/usr/local/lib/python3.10/contextlib.py:78)
Compilation Analysis:   <module> (/workspaces/dk4/pytorch/xla/test.py:34)
Compilation Analysis: --------------------------------------------------------------------------------
Compilation Analysis: ================================================================================

Post Compilation Analysis: ================================================================================
Post Compilation Analysis: Graph input size: 0.000001 GB
Post Compilation Analysis: Graph output size: 0.000115 GB
Post Compilation Analysis: Aliased Input size: 0.000000 GB
Post Compilation Analysis: Intermediate tensor size: 0.000000 GB
Post Compilation Analysis: Compiled program size: 0.000132 GB
Post Compilation Analysis: --------------------------------------------------------------------------------
Post Compilation Analysis: ================================================================================

Execution Analysis: ================================================================================
Execution Analysis: Execution Cause
Execution Analysis:   torch_xla.compile clear the pending graph prior calling the target function
Execution Analysis: Graph Info: 
Execution Analysis:   Graph Hash: 1abebf4362b8342721a64502bd4074c1
Execution Analysis:   Number of Graph Inputs: 1
Execution Analysis:   Number of Graph Outputs: 1
Execution Analysis: Python Frame Triggered Execution: 
Execution Analysis:   mark_step (/workspaces/dk4/pytorch/xla/torch_xla/core/xla_model.py:1008)
Execution Analysis:   _compile (/workspaces/dk4/pytorch/xla/torch_xla/torch_xla.py:108)
Execution Analysis:   __enter__ (/usr/local/lib/python3.10/contextlib.py:135)
Execution Analysis:   inner (/usr/local/lib/python3.10/contextlib.py:78)
Execution Analysis:   <module> (/workspaces/dk4/pytorch/xla/test.py:34)
Execution Analysis: --------------------------------------------------------------------------------
Execution Analysis: ================================================================================

Compilation Analysis: ================================================================================
Compilation Analysis: Compilation Cause
Compilation Analysis:   torch_xla.compile
Compilation Analysis: Graph Info: 
Compilation Analysis:   Graph Hash: 652d43861b869ad752e42502c3665a8b
Compilation Analysis:   Number of Graph Inputs: 3
Compilation Analysis:   Number of Graph Outputs: 2
Compilation Analysis: Python Frame Triggered Execution: 
Compilation Analysis:   mark_step (/workspaces/dk4/pytorch/xla/torch_xla/core/xla_model.py:1008)
Compilation Analysis:   sync (/workspaces/dk4/pytorch/xla/torch_xla/torch_xla.py:61)
Compilation Analysis:   _compile (/workspaces/dk4/pytorch/xla/torch_xla/torch_xla.py:115)
Compilation Analysis:   __exit__ (/usr/local/lib/python3.10/contextlib.py:142)
Compilation Analysis:   inner (/usr/local/lib/python3.10/contextlib.py:78)
Compilation Analysis:   <module> (/workspaces/dk4/pytorch/xla/test.py:34)
Compilation Analysis: --------------------------------------------------------------------------------
Compilation Analysis: ================================================================================

Post Compilation Analysis: ================================================================================
Post Compilation Analysis: Graph input size: 0.000127 GB
Post Compilation Analysis: Graph output size: 0.001717 GB
Post Compilation Analysis: Aliased Input size: 0.000000 GB
Post Compilation Analysis: Intermediate tensor size: 0.000000 GB
Post Compilation Analysis: Compiled program size: 0.000540 GB
Post Compilation Analysis: --------------------------------------------------------------------------------
Post Compilation Analysis: ================================================================================

Execution Analysis: ================================================================================
Execution Analysis: Execution Cause
Execution Analysis:   torch_xla.compile
Execution Analysis: Graph Info: 
Execution Analysis:   Graph Hash: 652d43861b869ad752e42502c3665a8b
Execution Analysis:   Number of Graph Inputs: 3
Execution Analysis:   Number of Graph Outputs: 2
Execution Analysis: Python Frame Triggered Execution: 
Execution Analysis:   mark_step (/workspaces/dk4/pytorch/xla/torch_xla/core/xla_model.py:1008)
Execution Analysis:   sync (/workspaces/dk4/pytorch/xla/torch_xla/torch_xla.py:61)
Execution Analysis:   _compile (/workspaces/dk4/pytorch/xla/torch_xla/torch_xla.py:115)
Execution Analysis:   __exit__ (/usr/local/lib/python3.10/contextlib.py:142)
Execution Analysis:   inner (/usr/local/lib/python3.10/contextlib.py:78)
Execution Analysis:   <module> (/workspaces/dk4/pytorch/xla/test.py:34)
Execution Analysis: --------------------------------------------------------------------------------
Execution Analysis: ================================================================================
```

### CODE REVIEW

**Comment by JackCaoG** [comment]
Test failure is irrelvant, I will rebase but this one should be ready for review.

File: torch_xla/torch_xla.py [hunk 1]
```diff
@@ -101,19 +101,21 @@ def foo2(x):
   """
 
   @contextlib.contextmanager
-  def _step():
+  def _compile():
     saved_eager_mode_status = torch_xla._XLAC._get_use_eager_mode()
     torch_xla._XLAC._set_use_eager_mode(False)
     # Clear pending operations
-    sync()
+    xm.mark_step()
```
Comment (will-cromar) [inline torch_xla/torch_xla.py:108]: Why is this changed to `mark_step`?

Comment (JackCaoG) [reply torch_xla/torch_xla.py:108]: there are two `sync` in the `torch_xla.compile`, one before traicng and one after tracing. In the C++ layer I am just looking at the python stack trace and it is not easy to figure out which one is which. I think I have 3 options

1. wrap one of the `sync` in to a function like `sync_graph_before_compile`
2. look at line number
3. use `mark_step` for one and `sync` for the other

I don't like first two so I ended up using the last one.

**Comment by JackCaoG** [comment]
Also I felt like right now `PT_XLA_DEBUG` frame output is a bit polluted because `sync` calls `mark_step`. I am thinking when users call `sync` we just remove the `mark_step` frame from the output. @will-cromar what do you think?

File: torch_xla/torch_xla.py (hunk 1 above)
Comment (will-cromar) [reply torch_xla/torch_xla.py:108]: Hahaha that's hacky

(1) is definitely the way to go here just to make clear what you're trying to do, and make this output more immediately readable to anyone else looking at the stack trace; there's not anything about `mark_step` that clearly communicates "before". You can add inner functions to avoid polluting the global namespace:

```
def compile

  def sync_before(): # or more explicitly, `clear_pending_ops`
      sync()

  def sync_after():
     sync()
```

Regarding the depth of the stack trace, we can easily flatten it a level by having `sync` just call `torch_xla._XLAC._xla_step_marker` directly instead of calling mark_step.

**Comment by JackCaoG** [comment]
@will-cromar can you take another look?

File: torch_xla/torch_xla.py [hunk 2]
```diff
@@ -56,9 +56,20 @@ def device_count() -> int:
...
+  """Launches all pending graph operations.
+  
+  Args:
+    wait (bool): whether to block the current process until the execution finished.
+
+  """
+  torch_xla._XLAC._xla_step_marker(
+      torch_xla._XLAC._xla_get_default_device(),
+      [],
+      wait=wait,
```
Comment (will-cromar) [inline torch_xla/torch_xla.py:69]: Isn't `wait` currently broken? Does it actually `WaitDeviceOps`?

File: torch_xla/torch_xla.py [hunk 3]
```diff
@@ -100,20 +111,24 @@ def foo2(x):
         res = foo2(x)
   """
 
+  def _clear_pending_ops_before_compile():
```
Comment (will-cromar) [inline torch_xla/torch_xla.py:114]: Thanks. Long-term, we should be able to remove this. If we assume eager execution outside of `compile`, then there should be no "pending ops"

**Review by will-cromar — APPROVED** [review]

File: torch_xla/torch_xla.py (hunk 2 above)
Comment (JackCaoG) [reply torch_xla/torch_xla.py:69]: let me check, if not let me fix in a follow up pr.

File: torch_xla/torch_xla.py (hunk 3 above)
Comment (JackCaoG) [reply torch_xla/torch_xla.py:114]: yea with eager there shouldn't be any pending ops.
